lib/data_magic/document_builder.rb

```diff
@@ -0,0 +1,140 @@
+      # fields: column_name => field_name
+      # config: DataMagic.Config instance for dictionary, column types, NULL
+      def parse_row(row, fields, config, options={}, additional=nil)
+        row = csv_row = row.to_hash
+        row = map_field_names(row, fields, options) unless fields.empty?
+        row = row.merge(calculated_fields(csv_row, config))
+        unless config.column_field_types.empty? && config.null_value.empty?
+          row = map_field_types(row, config.valid_types,
+                                config.column_field_types,
+                                config.null_value)
+        end
+        row = row.merge(additional) if additional
+        doc = NestedHash.new.add(row)
+        doc = parse_nested(doc, options) if options[:nest]
+        doc = select_only_fields(doc, options[:only]) unless options[:only].nil?
+        doc
+      end
+
+      def calculated_fields(row, config)
+        result = {}
+        config.calculated_field_list.each do |name|
+          result[name] = calculate(name, row, config.dictionary)
+        end
+        result
+      end
+
+      private
+
+      # row: a hash  (keys may be strings or symbols)
+      # valid_types: an array of allowed types
+      # field_types: hash field_name : type (float, integer, string)
+      # returns a hash where values have been coerced to the new type
+      def map_field_types(row, valid_types, field_types = {}, null_value = 'NULL')
+        mapped = {}
+        row.each do |key, value|
+          if value == null_value
+            mapped[key] = nil
+          else
+            type = field_types[key.to_sym] || field_types[key.to_s]
+            if valid_types.include? type
+              mapped[key] = fix_field_type(type, value, key)
+              mapped["_#{key}"] = value.downcase if type == "name"
+            else
+              fail InvalidDictionary, "unexpected type '#{type.inspect}' for field '#{key}'"
+            end
+          end
+        end
+        mapped
+      end
+
+      def parse_nested(document, options)
+        new_doc = {}
+        nest_options = options[:nest]
+        if nest_options
+          key = nest_options['key']
+          new_doc[key] = {}
+          new_doc['id'] = document['id'] unless document['id'].nil?
+          nest_options['contents'].each do |item_key|
+            new_doc[key][item_key] = document[item_key]
+          end
+        end
+        new_doc
+      end
+
+      def fix_field_type(type, value, key=nil)
+        return value if value.nil?
+
+        new_value = case type
+                    when "float"
+                      value.to_f
+                    when "integer"
+                      value.to_i
+                    when "lowercase_name"
+                      value.to_s.downcase   # used for searching
+                    else # "string"
+                      value.to_s
+        end
+        new_value = value.to_f if key and key.to_s.include? "location"
```

Comment:
This is not the PR to do it in, but we really need a better solution for lat-long floats than this, because matching on a word in the name is begging for problems.